[DOC] Instructions for oauth 2.0 token based authentication (#2013)
strimzi/strimzi-kafka-operator: documentation/book/oauth/proc-oauth-client-config.adoc

// Module included in the following assemblies:
//
// assembly-oauth-strimzi-config.adoc

[id='proc-oauth-client-config-{context}']

> = Configuring Kafka clients to use {oauth}

Reviewer: ```suggestion
= Configuring Kafka Java clients to use {oauth}
```

> This procedure describes how to configure Java-based Kafka clients to use {oauth} for interaction with Kafka brokers.

Reviewer: ```suggestion
This procedure describes how to configure Kafka Producer and Consumer APIs to use {oauth} for interaction with Kafka brokers.
```

Add a client callback plugin to your _pom.xml_ file, and configure the system properties.

.Prerequisites

* AMQ Streams and Kafka is running
* {oauth} authorization server is deployed and configured for OAuth access to Kafka brokers
* Kafka brokers are configured for {oauth}

.Procedure

. Add the client callback plugin to the `pom.xml` file for the Kafka client:
+
[source,xml]
----
<dependency>
 <groupId>io.strimzi</groupId>
 <artifactId>kafka-oauth-client</artifactId>

>  <version>0.1.0.redhat-00001</version>

Reviewer: We need to use the upstream version here.

```suggestion
 <version>0.1.0</version>
```

Author: Sorry. Missed that. I've set a variable for upstream and downstream.

</dependency>
----

. Configure the system properties for the callback:
+
[source,xml]
----

> System.setProperty(ClientConfig.OAUTH_TOKEN_ENDPOINT_URI, “http://<authorization-server-address>.nip.io/auth/realms/demo/protocol/openid-connect/token”);

Reviewer: ```suggestion
System.setProperty(ClientConfig.OAUTH_TOKEN_ENDPOINT_URI, “https://<authorization-server-address>/auth/realms/demo/protocol/openid-connect/token”);
```

System.setProperty(Config.OAUTH_CLIENT_ID, "kafka-producer-client");

> System.setProperty(Config.OAUTH_CLIENT_SECRET, "kafka-producer-client-secret");

Reviewer: ```suggestion
System.setProperty(Config.OAUTH_CLIENT_SECRET, "<client-secret>");
```

----

> . Enable SASL OAUTHBEARER and TLS in the Kafka client configuration:

Reviewer: ```suggestion
. Enable SASL OAUTHBEARER on TLS encrypted connection in the Kafka client configuration:
```

// Module included in the following module:
//
// con-oauth-config.adoc

[id='proc-oauth-client-config-{context}']
= Configuring Kafka Java clients to use {oauth}

This procedure describes how to configure Kafka producer and consumer APIs to use {oauth} for interaction with Kafka brokers.

Add a client callback plugin to your _pom.xml_ file, and configure the system properties.

.Prerequisites

* AMQ Streams and Kafka is running
* {oauth} authorization server is deployed and configured for OAuth access to Kafka brokers
* Kafka brokers are configured for {oauth}

.Procedure

. Add the client callback plugin to the `pom.xml` file for the Kafka client:
+
[source,xml,subs="+attributes"]
----
<dependency>
 <groupId>io.strimzi</groupId>
 <artifactId>kafka-oauth-client</artifactId>
 <version>{oauth-artifact-version}</version>
</dependency>
----

. Configure the system properties for the callback:
+
For example:
+
[source,xml, subs="+quotes,attributes"]
----
System.setProperty(ClientConfig.OAUTH_TOKEN_ENDPOINT_URI, “_https://<authorization-server-address>/auth/realms/demo/protocol/openid-connect/token_”); <1>
System.setProperty(Config.OAUTH_CLIENT_ID, "_<client-name>_"); <2>

> System.setProperty(Config.OAUTH_CLIENT_SECRET, "_<client-secret>_"); <3>

Reviewer: I think we can use `ClientConfig` for both of these to make the code a bit nicer (@mstruk The `ClientConfig` extends `Config`, right?)

Reviewer 2: Right. It would indeed make it look nicer.

Author: Updated

----
<1> URI of the authorization server token endpoint
<2> Client ID, which is the name used when creating the _client_ in the authorization server.
<3> Client secret created when creating the _client_ in the authorization server.

. Enable the _SASL OAUTHBEARER_ mechanism on a TLS encrypted connection in the Kafka client configuration:
+
[source,xml]
----
props.put("sasl.jaas.config", "org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required;");

> props.put("security.protocol", "SASL_SSL");

Author: I've added a callout.

// Module included in the following module:
//
// con-oauth-config.adoc

[id='proc-oauth-client-config-{context}']
= Configuring Kafka Java clients to use {oauth}

This procedure describes how to configure Kafka producer and consumer APIs to use {oauth} for interaction with Kafka brokers.

Add a client callback plugin to your _pom.xml_ file, and configure the system properties.

.Prerequisites

> * AMQ Streams and Kafka is running

Reviewer 3: Use the {ProductName} attribute.

is running > are running

> * {oauth} authorization server is deployed and configured for OAuth access to Kafka brokers

Reviewer 3: {oauth} authorization server > An {oauth} authorization server

* Kafka brokers are configured for {oauth}

.Procedure

. Add the client callback plugin to the `pom.xml` file for the Kafka client:
+
[source,xml,subs="+attributes"]
----
<dependency>
 <groupId>io.strimzi</groupId>
 <artifactId>kafka-oauth-client</artifactId>
 <version>{oauth-artifact-version}</version>
</dependency>
----

. Configure the system properties for the callback:
+
For example:
+

> [source,xml, subs="+quotes,attributes"]

Reviewer 3: The source is Java?

Author: It's the pom.xml - what do we normally use with such files?

----
System.setProperty(ClientConfig.OAUTH_TOKEN_ENDPOINT_URI, “_https://<authorization-server-address>/auth/realms/demo/protocol/openid-connect/token_”); <1>
System.setProperty(ClientConfig.OAUTH_CLIENT_ID, "_<client-name>_"); <2>
System.setProperty(ClientConfig.OAUTH_CLIENT_SECRET, "_<client-secret>_"); <3>
----

> <1> URI of the authorization server token endpoint

Reviewer 3: Missing full stop.
